REPO_LAYOUT.md

```diff
@@ -0,0 +1,88 @@
+as well as to clearly specify how extensions are added to the repository. The top level directories
+are:
+
+* [.circleci/](.circleci/): Configuration for [CircleCI](https://circleci.com/gh/envoyproxy).
+* [bazel/](bazel/): Configuration for Envoy's use of [Bazel](https://bazel.build/).
+* [ci/](ci/): Scripts used both during CI as well as to build Docker containers.
+* [configs/](configs/): Example Envoy configurations.
+* [docs/](docs/): Project level documentation is well as scripts for publishing final docs during
+  releases.
+* [examples/](examples/): Larger Envoy examples using Docker and Docker Compose.
+* [include/](include/): "Public" interface headers for "core" Envoy (not extensions). In general,
+  these are almost entirely 100% abstract classes. There are a few cases of not-abstract classes in
+  the "public" headers, typically for performance reasons.
+* [restarter/](restarter/): Envoy's hot restart wrapper Python script.
+* [source/](source/): Source code for core Envoy as well as extensions. The layout of this directory
+  is discussed in further detail below.
+* [support/](support/): Development support scripts (pre-commit Git hooks, etc.)
+* [test/](test/): Test code for core Envoy as well as extensions. The layout of this directory is
+  discussed in further detail below.
+* [tools/](tools/): Miscellaneous tools that have not found a home somewhere else.
+
+## [source/](source/)
+
+* [common/](source/[common/): Core Envoy code (not specific to extensions) that is also not
+  specific to a standalone server implementation. I.e., this is code that could be used if Envoy
+  were eventually embedded as a library.
+* [docs/](source/docs/): Miscellaneous developer/design documentation that is not relevant for
+  the public user documentation.
+* [exe/](source/exe/): Code specific to building the final production Envoy server binary. This is
+  the only code that is not shared by integration and unit tests.
+* [extensions/](source/extensions/): Extensions to the core Envoy code. The layout of this
+  directory is discussed in further detail below.
+* [server/](source/server/): Code specific to running Envoy as a standalone server. E.g,
+  configuration, server startup, workers, etc. Overtime, the line between `common/` and `server/`
+  has become somewhat blurred. Use best judgment as to where to place something.
+
+## [test/](test/)
+
+Not every directory within test is described below, but a few highlights:
+
+* Unit tests are found in directories matching their [source/](source/) equivalents. E.g.,
+  [common/](test/common/), [exe/](test/exe/), and [server/](test/server/).
+* Extension unit tests also match their source equivalents in [extensions/](test/extensions/).
+* [integration/](test/integration/) holds end-to-end integration tests using roughly the real
+  Envoy server code, fake downstream clients, and fake upstream servers. Integration tests also
+  test some of the extensions found in the repository.
```

Comment:
This doesn't seem ideal, aren't we going to push these into the extension specific dirs?

Reply:
This was my plan originally but @alyssawilk objected. Or maybe I misunderstood the objection @alyssawilk?

Reply:
We may - I'd asked we defer that until after the filter moves as I think there's a bunch of code which tests various filters, or tests core components (flow control) using extensions (various filters which buffer).

It's pretty non-trivial to tease these out.

Reply:
Also, we've been organizing our HTTP tests around codec so you can "easily" add a cors test for HTTP/2, HTTP, HTTP/2 upstream etc.  I think we need an integration test rewrite to make it easier to have "the cors filter test" which runs the test for the varions combinations and permutations, rather than having 3 integration tests per filter which would be really crummy from a scaling perspective.  I think if we do that refactor we could encourage folks to add new integration tests under their filter directories, and move the easy-to-tease-apart tests to their directories for reference implementations.  Still not sure how we want to handle testing core features which require filters without implementing a bunch of fake test filters (which we could totally do, but don't have yet) so I think this will be a series of TODOs and follow-up PRs.

Reply:
Yeah, I agree that more thinking is needed here. I would prefer that we not let "perfect be the enemy of the good" on this so I will add some aspirational text on where we would like this to go.